File: servidor.js
const http = require('http');
const fs = require('fs');
const path = require('path');

const PORT = process.env.PORT || 3000;

function obtenerDimensiones(matriz) {
    return {
        filas: matriz.length,
        columnas: matriz[0].length
    };
}

function esMatrizValida(matriz) {
    if (!Array.isArray(matriz) || matriz.length === 0 || !Array.isArray(matriz[0]) || matriz[0].length === 0) {
        return false;
    }

    const columnas = matriz[0].length;

    for (let i = 0; i < matriz.length; i++) {
        if (!Array.isArray(matriz[i]) || matriz[i].length !== columnas) {
            return false;
        }

        for (let j = 0; j < matriz[i].length; j++) {
            if (typeof matriz[i][j] !== 'number' || Number.isNaN(matriz[i][j])) {
                return false;
            }
        }
    }

    return true;
}

function ponerEncabezadosCors(res, contentType) {
    const headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type'
    };

    if (contentType) {
        headers['Content-Type'] = contentType;
    }

    res.writeHead(200, headers);
}

const servidor = http.createServer(function(req, res) {
    if (req.method === 'OPTIONS') {
        res.writeHead(204, {
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
            'Access-Control-Allow-Headers': 'Content-Type'
        });
        res.end();
        return;
    }

    if (req.method === 'GET' && req.url === '/') {
        const rutaIndex = path.join(__dirname, 'index.html');
        fs.readFile(rutaIndex, function(err, contenido) {
            if (err) {
                res.writeHead(500, {'Content-Type': 'text/plain; charset=utf-8'});
                res.end('Error al cargar index.html');
                return;
            }
            ponerEncabezadosCors(res, 'text/html; charset=utf-8');
            res.end(contenido);
        });
        return;
    }

    if (req.method === 'GET' && req.url === '/script.js') {
        const rutaScript = path.join(__dirname, 'script.js');
        fs.readFile(rutaScript, function(err, contenido) {
            if (err) {
                res.writeHead(500, {'Content-Type': 'text/plain; charset=utf-8'});
                res.end('Error al cargar script.js');
                return;
            }
            ponerEncabezadosCors(res, 'application/javascript; charset=utf-8');
            res.end(contenido);
        });
        return;
    }

    if (req.method === 'POST' && req.url === '/calcular') {
        let cuerpo = '';
        req.on('data', function(chunk) {
            cuerpo += chunk;
        });

        req.on('end', function() {
            let datos;
            try {
                datos = JSON.parse(cuerpo);
            } catch (e) {
                res.writeHead(400, {
                    'Access-Control-Allow-Origin': '*',
                    'Content-Type': 'application/json'
                });
                res.end(JSON.stringify({ error: 'JSON inválido' }));
                return;
            }

            const m1 = datos.matriz1;
            const m2 = datos.matriz2;
            const operacion = datos.operacion || 'suma';

            if (!esMatrizValida(m1) || !esMatrizValida(m2)) {
                res.writeHead(400, {
                    'Access-Control-Allow-Origin': '*',
                    'Content-Type': 'application/json'
                });
                res.end(JSON.stringify({ error: 'Matrices inválidas' }));
                return;
            }

            const resultado = [];
            const dim1 = obtenerDimensiones(m1);
            const dim2 = obtenerDimensiones(m2);

            if (operacion === 'multiplicacion') {
                if (dim1.columnas !== dim2.filas) {
                    res.writeHead(400, {
                        'Access-Control-Allow-Origin': '*',
                        'Content-Type': 'application/json'
                    });
                    res.end(JSON.stringify({ error: 'Dimensiones incompatibles para multiplicación' }));
                    return;
                }

                for (let i = 0; i < dim1.filas; i++) {
                    const filaMultiplicacion = [];
                    for (let j = 0; j < dim2.columnas; j++) {
                        let suma = 0;
                        for (let k = 0; k < dim2.filas; k++) {
                            suma = suma + (m1[i][k] * m2[k][j]);
                        }
                        filaMultiplicacion.push(suma);
                    }
                    resultado.push(filaMultiplicacion);
                }
            } else {
                if (dim1.filas !== dim2.filas || dim1.columnas !== dim2.columnas) {
                    res.writeHead(400, {
                        'Access-Control-Allow-Origin': '*',
                        'Content-Type': 'application/json'
                    });
                    res.end(JSON.stringify({ error: 'Dimensiones incompatibles para suma' }));
                    return;
                }

                for (let i = 0; i < dim1.filas; i++) {
                    const filaRes = [];
                    for (let j = 0; j < dim1.columnas; j++) {
                        filaRes.push(m1[i][j] + m2[i][j]);
                    }
                    resultado.push(filaRes);
                }
            }

            res.writeHead(200, {
                'Access-Control-Allow-Origin': '*',
                'Content-Type': 'application/json'
            });
            res.end(JSON.stringify({ resultado: resultado }));
        });
        return;
    }

    res.writeHead(404, {
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'text/plain; charset=utf-8'
    });
    res.end('Ruta no encontrada');
});

servidor.on('error', function(err) {
    if (err.code === 'EADDRINUSE') {
        console.error('El puerto ' + PORT + ' ya está en uso. Cerrá el proceso anterior o usá otro puerto.');
        process.exit(1);
    }

    console.error('Error del servidor:', err.message);
    process.exit(1);
});

servidor.listen(PORT, function() {
    console.log('Servidor corriendo en puerto ' + PORT);
});

File: servidor.py
from http.server import BaseHTTPRequestHandler, HTTPServer
import json
import os


PORT = 5000


class MiServidor(BaseHTTPRequestHandler):
    def _poner_encabezados_cors(self, codigo, tipo="text/plain; charset=utf-8"):
        self.send_response(codigo)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.send_header("Content-Type", tipo)
        self.end_headers()

    def _enviar_texto(self, codigo, texto, tipo="text/plain; charset=utf-8"):
        self._poner_encabezados_cors(codigo, tipo)
        self.wfile.write(texto.encode("utf-8"))

    def _cargar_archivo(self, nombre_archivo, tipo):
        try:
            with open(nombre_archivo, "rb") as archivo:
                contenido = archivo.read()
            self._poner_encabezados_cors(200, tipo)
            self.wfile.write(contenido)
        except OSError:
            self._enviar_texto(500, "No se pudo cargar el archivo")

    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.end_headers()

    def do_GET(self):
        if self.path == "/":
            self._cargar_archivo("index.html", "text/html; charset=utf-8")
            return

        if self.path == "/script.js":
            self._cargar_archivo("script.js", "application/javascript; charset=utf-8")
            return

        self._enviar_texto(404, "Ruta no encontrada")

    def do_POST(self):
        if self.path != "/calcular":
            self._enviar_texto(404, "Ruta no encontrada")
            return

        largo = int(self.headers.get("Content-Length", 0))
        cuerpo = self.rfile.read(largo).decode("utf-8")

        try:
            datos = json.loads(cuerpo)
        except json.JSONDecodeError:
            self._enviar_json(400, {"error": "JSON inválido"})
            return

        matriz1 = datos.get("matriz1")
        matriz2 = datos.get("matriz2")
        operacion = datos.get("operacion", "suma")

        if not isinstance(matriz1, list) or not isinstance(matriz2, list):
            self._enviar_json(400, {"error": "Matrices inválidas"})
            return

        if operacion == "multiplicacion":
            resultado = self._multiplicar(matriz1, matriz2)
            if resultado is None:
                self._enviar_json(400, {"error": "Dimensiones incompatibles para multiplicación"})
                return
        else:
            resultado = self._sumar(matriz1, matriz2)
            if resultado is None:
                self._enviar_json(400, {"error": "Dimensiones incompatibles para suma"})
                return

        self._enviar_json(200, {"resultado": resultado})

    def _enviar_json(self, codigo, datos):
        contenido = json.dumps(datos)
        self._poner_encabezados_cors(codigo, "application/json; charset=utf-8")
        self.wfile.write(contenido.encode("utf-8"))

    def _sumar(self, matriz1, matriz2):
        if len(matriz1) != len(matriz2):
            return None

        resultado = []
        for i in range(len(matriz1)):
            if not isinstance(matriz1[i], list) or not isinstance(matriz2[i], list):
                return None
            if len(matriz1[i]) != len(matriz2[i]):
                return None

            fila = []
            for j in range(len(matriz1[i])):
                fila.append(matriz1[i][j] + matriz2[i][j])
            resultado.append(fila)

        return resultado

    def _multiplicar(self, matriz1, matriz2):
        if len(matriz1) == 0 or len(matriz2) == 0:
            return None
        if not isinstance(matriz1[0], list) or not isinstance(matriz2[0], list):
            return None
        if len(matriz1[0]) != len(matriz2):
            return None

        resultado = []
        for i in range(len(matriz1)):
            fila = []
            for j in range(len(matriz2[0])):
                suma = 0
                for k in range(len(matriz2)):
                    suma = suma + (matriz1[i][k] * matriz2[k][j])
                fila.append(suma)
            resultado.append(fila)

        return resultado


def main():
    servidor = HTTPServer(("localhost", PORT), MiServidor)
    print("Servidor corriendo en puerto " + str(PORT))
    servidor.serve_forever()


if __name__ == "__main__":
    main()
